# Collecting Corpus Data for Thesis

For the goal of our thesis we need a representative corpus of news documents

In [1]:
import pandas as pd
from datasets import load_dataset

/Users/i.mrasov/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


## 1. Collect Data Before 2019 (yutkin Kaggle repo, 1999-2019)

Source: [Kaggle Dataset](https://www.kaggle.com/datasets/yutkin/corpus-of-russian-news-articles-from-lenta)

In [2]:
df_before_2019 = pd.read_csv("/Users/i.mrasov/Downloads/lenta-ru-news.csv")

/var/folders/47/3f9lnsqj7_dc3k8ybjy6cx240000gp/T/ipykernel_55882/611430653.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df_before_2019 = pd.read_csv("/Users/i.mrasov/Downloads/lenta-ru-news.csv")


In [12]:
print('Dataset shape:', df_before_2019.shape)
print('Columns:', df_before_2019.columns)
df_before_2019.head()

Dataset shape: (800975, 6)
Columns: Index(['url', 'title', 'text', 'topic', 'tags', 'date'], dtype='object')


,url,title,text,topic,tags,date
0,https://lenta.ru/news/1914/09/16/hungarnn/,1914. Русские войска вступили в пределы Венгрии,Бои у Сопоцкина и Друскеник закончились отступ...,Библиотека,Первая мировая,1914/09/16
1,https://lenta.ru/news/1914/09/16/lermontov/,1914. Празднование столетия М.Ю. Лермонтова от...,"Министерство народного просвещения, в виду про...",Библиотека,Первая мировая,1914/09/16
2,https://lenta.ru/news/1914/09/17/nesteroff/,1914. Das ist Nesteroff!,"Штабс-капитан П. Н. Нестеров на днях, увидев в...",Библиотека,Первая мировая,1914/09/17
3,https://lenta.ru/news/1914/09/17/bulldogn/,1914. Бульдог-гонец под Льежем,Фотограф-корреспондент Daily Mirror рассказыва...,Библиотека,Первая мировая,1914/09/17
4,https://lenta.ru/news/1914/09/18/zver/,1914. Под Люблином пойман швабский зверь,"Лица, приехавшие в Варшаву из Люблина, передаю...",Библиотека,Первая мировая,1914/09/18


We are working with large corpus so to ensure the domain-specific approach we would need to filter data based on the `topic` values

In [7]:
print(df_before_2019['topic'].value_counts())

topic
Россия               160445
Мир                  136621
Экономика             79528
Спорт                 64413
Культура              53797
Бывший СССР           53402
Наука и техника       53136
Интернет и СМИ        44663
Из жизни              27605
Дом                   21734
Силовые структуры     19596
Ценности               7766
Бизнес                 7399
Путешествия            6408
69-я параллель         1268
Крым                    666
Культпросвет            340
Легпром                 114
Библиотека               65
Оружие                    3
ЧМ-2014                   2
МедНовости                1
Сочи                      1
Name: count, dtype: int64


Initial choices:

- Экономика             79528

- Бизнес                 7399

Let's filter news articles according to these topics. This way we can ensure domain specificity of news and focus on socio-economic documents

In [16]:
df_before_2019_economic = df_before_2019[(df_before_2019['topic'] == 'Экономика') | (df_before_2019['topic'] == 'Бизнес')]
print(df_before_2019_economic.shape)

(86927, 6)


In [17]:
df_before_2019_economic.head()

,url,title,text,topic,tags,date
630,https://lenta.ru/news/1999/10/04/tv/,Телеканалы станут вещать по единому тарифу,С 1 января 2000 года все телеканалы будут опла...,Экономика,Все,1999/10/04
635,https://lenta.ru/news/1999/10/04/volkswagen/,"Volkswagen выкупает остатки акций ""Шкоды""",Германский автопромышленный концерн Volkswagen...,Экономика,Все,1999/10/04
636,https://lenta.ru/news/1999/10/04/tumen/,Прибыль Тюменнефтегаза возросла в 10 раз,"Нераспределенная прибыль ОАО ""Тюменнефтегаз"", ...",Экономика,Все,1999/10/04
660,https://lenta.ru/news/1999/10/05/sprint/,Крупнейшее в истории слияние компаний происход...,Две крупнейших телекоммуникационных компании С...,Экономика,Все,1999/10/05
661,https://lenta.ru/news/1999/10/05/volga/,ГАЗ получил четверть обещанного кредита,"ОАО ""ГАЗ"" и Нижегородский банк Сбербанка Росси...",Экономика,Все,1999/10/05


Checking for data — deleting duplicates and empty rows

In [19]:
df_before_2019_economic.isna().sum()

url         0
title       0
text        0
topic       0
tags     3129
date        0
dtype: int64

In [24]:
print(df_before_2019_economic.duplicated().sum())
print(df_before_2019_economic.duplicated(subset=["url"]).sum())

np.int64(0)

Nice, no empty rows and duplicates. Let's convert data to correct dtypes

In [27]:
df_before_2019_economic.dtypes

url              object
title            object
text             object
topic            object
tags             object
date     datetime64[ns]
dtype: object

In [20]:
df_before_2019_economic["date"] = pd.to_datetime(df_before_2019_economic["date"], errors="coerce")

/var/folders/47/3f9lnsqj7_dc3k8ybjy6cx240000gp/T/ipykernel_55882/3415735303.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_before_2019_economic["date"] = pd.to_datetime(df_before_2019_economic["date"], errors="coerce")


In [21]:
df_before_2019_economic.dtypes

url              object
title            object
text             object
topic            object
tags             object
date     datetime64[ns]
dtype: object

## 2. Collect Data After 2019 (data-silence HuggingFace, 2019-2024)

Source: [HuggingFace](https://huggingface.co/datasets/data-silence/lenta.ru_2-extended)

In [15]:
df_from_2019 = pd.read_csv("hf://datasets/data-silence/lenta.ru_2-extended/final_lenta_new.csv.gz")

KeyboardInterrupt: 

In [28]:
print(df_from_2019.shape)
print(df_from_2019.columns)
df_from_2019.head()

(577134, 7)
Index(['url', 'date', 'title', 'topic', 'news', 'tags', 'links'], dtype='object')


,url,date,title,topic,news,tags,links
0,https://lenta.ru/news/2024/04/02/v-kurskoy-obl...,2024-04-02 21:16:00,В Курской области объявили опасность атаки БПЛА,Россия,В Курской области объявили опасность атаки бес...,Происшествия,"['https://t.me/gubernator_46/5361', 'https://l..."
1,https://lenta.ru/news/2024/04/02/yuristy-proko...,2024-04-02 22:39:00,Юристы прокомментировали иск пострадавших в от...,Силовые структуры,"Владельцам «Крокус Сити Холла», в котором 22 м...",Следствие и суд,"['http://ria.ru/', 'https://lenta.ru/news/2024..."
2,https://lenta.ru/news/2024/04/02/na-zapade-ras...,2024-04-02 18:28:00,На Западе рассказали о нехватке Patriot для пе...,Мир,"Военный аналитик Александр Меркурис рассказал,...",Политика,['https://www.youtube.com/watch?v=vra-b_6tAI0'...
3,https://lenta.ru/news/2024/04/02/ocherednye-ro...,2024-04-02 17:45:00,Очередные российские спортсмены отказались от ...,Спорт,Российские гребцы отказались от участия в квал...,Летние виды,['https://rowingrussia.ru/news/rossiyskie-greb...
4,https://lenta.ru/news/2024/04/03/bayden-i-si-t...,2024-04-03 02:03:00,Байден и Си Цзиньпин обсудили ситуацию вокруг ...,Мир,Президент США Джо Байден и председатель Китая ...,Политика,['http://tass.ru/']


In [29]:
print(df_from_2019['topic'].value_counts())

topic
Россия               111261
Мир                   92838
Бывший СССР           61837
Экономика             57548
Силовые структуры     37741
Спорт                 31909
Наука и техника       29044
Интернет и СМИ        25018
Среда обитания        22272
Путешествия           22156
Ценности              21863
Культура              20591
Из жизни              19794
Забота о себе          9550
Моя страна             8181
Нацпроекты             2978
69-я параллель         2149
Дом                     255
Бизнес                   37
Оружие                   11
Мотор                     2
Name: count, dtype: int64


Here we choose once again these topics:

- Экономика             57548

Бизнес is not included, only has 37 articles (insignificant)

In [30]:
df_from_2019_economic = df_from_2019[df_from_2019['topic'] == 'Экономика']
print(df_from_2019_economic.shape)

(57548, 7)


In [32]:
df_from_2019_economic.head()

,url,date,title,topic,news,tags,links
12,https://lenta.ru/news/2024/04/02/nazvano-chisl...,2024-04-02 19:49:00,Названо число россиян с кредитами,Экономика,По данным на 1 января 2024 года кредиты в Росс...,Капитал,"['https://www.cbr.ru/press/event/?id=18579', '..."
35,https://lenta.ru/news/2024/04/03/rossiyane-vos...,2024-04-03 08:55:00,Россияне вошли в топ-10 любителей импортного а...,Экономика,В 2023 году Россия вошла в топ-10 и заняла дев...,Бизнес,"['http://ria.ru/', 'https://lenta.ru/news/2024..."
60,https://lenta.ru/news/2024/04/03/currency/,2024-04-03 10:14:00,Россиянам дали совет по покупке валюты,Экономика,Россиянам перед покупкой валюты в ближайшее вр...,Капитал,['https://ru.tradingview.com/symbols/USDRUB/?e...
65,https://lenta.ru/news/2024/04/03/nedobro/,2024-04-03 08:35:00,Российские прокуроры объяснили изъятие акций у...,Экономика,Все владельцы акций Соликамского магниевого за...,Бизнес,['https://www.rbc.ru/finances/03/04/2024/660c0...
66,https://lenta.ru/news/2024/04/03/rekordnye-pro...,2024-04-03 11:54:00,Рекордные продажи новых легковушек в России об...,Экономика,В последнюю неделю марта продажи новых легковы...,Бизнес,"['https://www.kommersant.ru/doc/6612081', 'htt..."


Checking for data — deleting duplicates and empty rows

In [33]:
df_from_2019_economic.isna().sum()

url         0
date        0
title       0
topic       0
news        0
tags     1974
links       0
dtype: int64

In [34]:
print(df_from_2019_economic.duplicated().sum())
print(df_from_2019_economic.duplicated(subset=["url"]).sum())

0
0


Once again, no empty rows and duplicates. Let's convert data to correct dtypes

In [35]:
df_from_2019_economic.dtypes

url      object
date     object
title    object
topic    object
news     object
tags     object
links    object
dtype: object

In [36]:
df_from_2019_economic["date"] = pd.to_datetime(df_from_2019_economic["date"], errors="coerce")

/var/folders/47/3f9lnsqj7_dc3k8ybjy6cx240000gp/T/ipykernel_55882/3712253233.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_from_2019_economic["date"] = pd.to_datetime(df_from_2019_economic["date"], errors="coerce")


In [37]:
df_from_2019_economic.dtypes

url              object
date     datetime64[ns]
title            object
topic            object
news             object
tags             object
links            object
dtype: object

Also, let's delete `links` column as it is not needed for topic modeling

In [ ]:
df_from_2019_economic = df_from_2019_economic.drop(columns=['links'])
df_from_2019_economic.head()

And to concatenate the latter dataset with the former we need to reorder the columns 

In [47]:
print('Structure before 2019:', df_before_2019_economic.columns)
print('Structure from 2019:', df_from_2019_economic.columns)

Structure before 2019: Index(['url', 'title', 'text', 'topic', 'tags', 'date'], dtype='object')
Structure from 2019: Index(['url', 'date', 'title', 'topic', 'news', 'tags'], dtype='object')


In [ ]:
df_from_2019_economic = df_from_2019_economic[['url', 'title', 'news', 'topic', 'tags', 'date']]
df_from_2019_economic = df_from_2019_economic.rename(columns={'news': 'text'})
df_from_2019_economic.head()

## 3. Concatenating Two Datasets 

In [51]:
df_combined = pd.concat([df_before_2019_economic, df_from_2019_economic], ignore_index=True)
print('Full corpus shape:', df_combined.shape)
df_combined.head()

Full corpus shape: (144475, 6)


,url,title,text,topic,tags,date
0,https://lenta.ru/news/1999/10/04/tv/,Телеканалы станут вещать по единому тарифу,С 1 января 2000 года все телеканалы будут опла...,Экономика,Все,1999-10-04
1,https://lenta.ru/news/1999/10/04/volkswagen/,"Volkswagen выкупает остатки акций ""Шкоды""",Германский автопромышленный концерн Volkswagen...,Экономика,Все,1999-10-04
2,https://lenta.ru/news/1999/10/04/tumen/,Прибыль Тюменнефтегаза возросла в 10 раз,"Нераспределенная прибыль ОАО ""Тюменнефтегаз"", ...",Экономика,Все,1999-10-04
3,https://lenta.ru/news/1999/10/05/sprint/,Крупнейшее в истории слияние компаний происход...,Две крупнейших телекоммуникационных компании С...,Экономика,Все,1999-10-05
4,https://lenta.ru/news/1999/10/05/volga/,ГАЗ получил четверть обещанного кредита,"ОАО ""ГАЗ"" и Нижегородский банк Сбербанка Росси...",Экономика,Все,1999-10-05


One important check for duplicates in a combined datasets 

In [52]:
print(df_combined.duplicated(subset=["url"]).sum())

0


This is our full corpus for the work. Store it in .csv for now

In [53]:
df_combined.to_csv("lenta_socioeconomic_corpus.csv", index=False)